### PR TITLE
Fixes #8783 , cargo new fails without a author name or email

### LINKED ISSUE #8783: Cargo new fails with side effect when $USER env variable is not set
**Problem**
When running `cargo new foo` in the vscode built-in terminal it fails with error `$USER` env variable is not set but it does so after initializing the package repository `foo` so now I have repo created but it is not initialized

```zsh
└── foo
    ├── .git
    └── .gitignore
```

The expectation is if the command fails there should be no side effect, no folder or git repo created.

**Steps**

1. `unset USER`
2. `cargo new foo`
3.  `ls`

**Possible Solution(s)**
[discover_author](https://github.com/rust-lang/cargo/blob/2fbce89dd625bcd49ce7fd75f02d793ce0e1b658/src/cargo/ops/cargo_new.rs#L648) bails when `$USER` is not found, so it should be called before [init_vcs](https://github.com/rust-lang/cargo/blob/2fbce89dd625bcd49ce7fd75f02d793ce0e1b658/src/cargo/ops/cargo_new.rs#L645)

**Notes**

Output of `cargo version`:
`cargo 1.47.0 (f3c7e066a 2020-08-28)`

### PR DESCRIPTION
**Pull request by aniljava**
If user can not be obtained from git or env variables $USER, new command defaults to empty author in generated Cargo.toml

Could not edit old PR(#8910 8783) as the original clone was deleted.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! Would you be able to add some tests for this? Additionally if we can't find the email or the username then I think we should probably set `authors = []` (or omit it entirely) in the manifest rather than setting `authors = [""]` which I think is what happens with this PR?

**Comment by aniljava** [comment]
Thank you for looking into it. I will update this PR with unit tests and omit authors from manifest if it is empty.

**Comment by aniljava** [comment]
Currently the behavior is to set authors = [] instead of omiting it.

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -645,19 +645,37 @@ fn mk(config: &Config, opts: &MkOptions<'_>) -> CargoResult<()> {
...
-        | (None, Some(email), name, _)
-        | (None, None, name, Some(email)) => {
+    let (discovered_name, discovered_email) = discover_author(path);
+
+    // "Name <email>" or "Name" or "<email>" or None if neither name nor email is obtained
+    // cfg takes priority over the discovered ones
+    let author_name = match (cfg.name, discovered_name) {
+        (Some(name), _) | (_, Some(name)) => Some(name),
+        (None, None) => None,
+    };
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs:655]: I think this and the below `match` can be expressed as `cfg.name.or(discovered_name)`

**Review by alexcrichton — COMMENTED** [review]
Looks good to me! Just one stylistic nit

File: src/cargo/ops/cargo_new.rs (hunk 1 above)
Comment (aniljava) [reply src/cargo/ops/cargo_new.rs]: Thank you for your feedback and guidance.

**Comment by alexcrichton** [comment]
Hm it looks like tests may be failing?

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**